Sources/PackageLoading/PackageBuilder.swift

```diff
@@ -1253,37 +1253,44 @@ public final class PackageBuilder {
             append(Product(name: product.name, type: product.type, targets: targets))
         }
 
-        // Add implicit executables - for root packages only.
+        // Add implicit executables - for root packages and for dependency plug‐ins.
 
-        if self.manifest.packageKind.isRoot {
-            // Compute the list of targets which are being used in an
-            // executable product so we don't create implicit executables
-            // for them.
-            let explicitProductsTargets = Set(self.manifest.products.flatMap{ product -> [String] in
-                switch product.type {
-                case .library, .plugin, .test:
-                    return []
-                case .executable, .snippet:
-                    return product.targets
-                }
-            })
-
-            let productMap = products.reduce(into: [String: Product]()) { partial, iterator in
-                partial[iterator.key] = iterator.item
+        // Compute the list of targets which are being used in an
+        // executable product so we don't create implicit executables
+        // for them.
+        let explicitProductsTargets = Set(self.manifest.products.flatMap{ product -> [String] in
+            switch product.type {
+            case .library, .plugin, .test:
+                return []
+            case .executable, .snippet:
+                return product.targets
             }
+        })
 
-            for target in targets where target.type == .executable {
-                if explicitProductsTargets.contains(target.name) {
-                    // If there is already an executable target with this name, skip generating a product for it
-                    continue
-                } else if let product = productMap[target.name] {
-                    // If there is already a product with this name skip generating a product for it,
-                    // but warn if that product is not executable
-                    if product.type != .executable {
-                        self.observabilityScope.emit(warning: "The target named '\(target.name)' was identified as an executable target but a non-executable product with this name already exists.")
-                    }
-                    continue
-                } else {
+        let productMap = products.reduce(into: [String: Product]()) { partial, iterator in
+            partial[iterator.key] = iterator.item
+        }
+
+        let implicitPlugInExecutables = Set(targets.lazy
+             .filter({ $0.type == .plugin })
+             .flatMap({ $0.dependencies })
+             .map { $0.name })
```

Comment:
If I read this correctly, this prevents non-plugin targets from depending on these implicit products (which would be great).  Or am I reading that wrong?

Reply:
Probably...? It certainly means they are not there to depend on during resolution or loading. But later when the build system is constructed there is no insulation to prevent dangling references from pointing where it will end up. What I do not know for sure, is if there is a way for it to sneak under the diagnostic checks to get that far, since some dangling references are temporarily permitted so that later diagnostics have more information available to suggest fixes.

Reply:
Thanks for clarifying — it's very possible that we should (in a separate PR, by me) add more checks for that kind of thing.

Representing plugins as targets definitely has some trade-offs.  It's a bit odd to do so, but making them completely separate would have made it more difficult to fit them into the dependency graph.  I still think it would be even more complicated to make them participate in the dependency graph without being targets, but it definitely does introduce some edge cases.